dev/python_script_generator/Generator - Design Document.md

```diff
@@ -0,0 +1,170 @@
+- **output_types** – The types of output in a multi output scenario. Default is None, and means every multi output is a matrix.
+
+
+
+Lets take the DML script of `kmeans` and its function definition as an example.
+
+```R
+m_kmeans = function(Matrix[Double] X, Integer k = 10, Integer runs = 10, Integer max_iter = 1000,
+                    Double eps = 0.000001, Boolean is_verbose = FALSE, Integer avg_sample_size_per_centroid = 50,
+    				Integer seed = -1)
+           return (Matrix[Double] C, Matrix[Double] Y)
+```
+
+Using the dml function definition we can create the `OperationNode` as follows:
+
+`Matrix[Double] X` indicates a required parameter with no default value. Since `X` is a `Matrix` we will specify the python input parameter type for `x` to be an `OperationNode`. Since all the other parameters are optional input parameters we will assign them to `‘**kwargs: Dict[str, VALID_INPUT_TYPES]`. The return type of the python function is `OperationNode` because an `OperationNode` object always returns an `OperationNode` object as defined in the python API documentation. Following this specification we would generate a python function definition that looks as follows:
+
+```python
+def kmeans(x: OperationNode, **kwargs: Dict[str, VALID_INPUT_TYPES]) -> OperationNode:
+```
+
+After the function definition the python script file structure contains a multi line-string comment with parameter information which we will parse from the dml script header.
+
+Subsequently some input parameter checks are performed. `x._check_matrix_op()` should we called on `OperationNode` typed inputs to ensure that it is a `matrix`. 
+
+In the next step the `OperationNode` requires a dictionary of the input parameters. For the required input parameters we have to assign a key which we can take from the dml script function definition. All optional parameters can simply be updated to the dictionary.
+
+The code below shows how a python function should look like when parsed from the dml script.
+
+````python
+def kmeans(x: OperationNode, **kwargs: Dict[str, VALID_INPUT_TYPES]) -> OperationNode:
+    """
+    Performs KMeans on matrix input.
+
+    :param x: Input dataset to perform K-Means on.
+    :param k: The number of centroids to use for the algorithm.
+    :param runs: The number of concurrent instances of K-Means to run (with different initial centroids).
+    :param max_iter: The maximum number of iterations to run the K-Means algorithm for.
+    :param eps: Tolerance for the algorithm to declare convergence using WCSS change ratio.
+    :param is_verbose: Boolean flag if the algorithm should be run in a verbose manner.
+    :param avg_sample_size_per_centroid: The average number of records per centroid in the data samples.
+    :return: `OperationNode` List containing two outputs 1. the clusters, 2 the cluster ID associated with each row in x.
```

Comment:
the multi return is not nice as it is, so if you have better suggestions feel free to design it better.